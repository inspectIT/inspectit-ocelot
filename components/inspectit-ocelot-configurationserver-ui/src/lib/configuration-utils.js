import { CONFIGURATION_TYPES } from '../data/constants';
import _ from 'lodash';

/**
 * Returns the type of the specified configuration. In case the configuration is null or empty
 * YAML will be returned.
 *
 * Example: the following configuration will result in type "METHOD_CONFIGURATION"
 * > # {"type": "METHOD_CONFIGURATION"}
 * > inspectit:
 * >   config:
 * >     ...
 *
 * @param {string} yamlConfiguration
 */
export const getConfigurationType = (yamlConfiguration) => {
  if (!yamlConfiguration) {
    // empty files will also result in YAML types.
    return CONFIGURATION_TYPES.YAML;
  }

  const lines = yamlConfiguration.split('\n');
  const headerLine = lines[0].trim();

  // special files must define a header comment
  if (!headerLine.startsWith('#')) {
    return CONFIGURATION_TYPES.YAML;
  }

  const jsonString = headerLine.substring(1);
  try {
    const { type } = JSON.parse(jsonString);
    return _.find(CONFIGURATION_TYPES, { name: type }, CONFIGURATION_TYPES.YAML);
  } catch {
    // YAML type in case the first comment is no JSON object
    return CONFIGURATION_TYPES.YAML;
  }
};
