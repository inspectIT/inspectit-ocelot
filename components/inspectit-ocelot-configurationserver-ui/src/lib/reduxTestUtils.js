import { Provider } from 'react-redux';
import React from 'react';
import { configureStore } from '@reduxjs/toolkit';

/**
 * wrapper function for testing purposes that include a redux store
 */
export function storeWrapper(jsx, reducers) {
  const store = mockStore(reducers);
  return <Provider store={store}>{jsx}</Provider>;
}

/**
 * store configuration function that supplies a store with the provided reducers
 */
export function mockStore(reducers) {
  return configureStore({
    reducer: reducers,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware({ serializableCheck: false }),
  });
}
