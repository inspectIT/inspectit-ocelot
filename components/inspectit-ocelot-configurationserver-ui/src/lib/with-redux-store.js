import React from 'react';
import configureStore from '../redux/store';

const isServer = typeof window === 'undefined';
const __NEXT_REDUX_STORE__ = '__NEXT_REDUX_STORE__';

function getOrCreateStore(initialState) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return configureStore(initialState, true);
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[__NEXT_REDUX_STORE__]) {
    window[__NEXT_REDUX_STORE__] = configureStore(initialState, false);
  }
  return window[__NEXT_REDUX_STORE__];
}

function getStore() {
  return window ? window[__NEXT_REDUX_STORE__] : null;
}

export { getStore };

const AppWithRedux = (App) => {
  return class AppWithRedux extends React.Component {
    static async getInitialProps(appContext) {
      // Get or Create the store with `undefined` as initialState
      // This allows you to set a custom default initialState
      const reduxStore = getOrCreateStore();

      // Provide the store to getInitialProps of pages
      appContext.ctx.reduxStore = reduxStore;

      let appProps = {};
      if (typeof App.getInitialProps === 'function') {
        appProps = await App.getInitialProps(appContext);
      }

      return {
        ...appProps,
        initialReduxState: reduxStore.getState(),
      };
    }

    constructor(props) {
      super(props);
      this.reduxStore = getOrCreateStore(props.initialReduxState);
    }

    render() {
      return <App {...this.props} reduxStore={this.reduxStore} />;
    }
  };
};
export default AppWithRedux;
