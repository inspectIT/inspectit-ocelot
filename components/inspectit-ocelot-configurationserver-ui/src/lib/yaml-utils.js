/**
 * A heuristic for getting the "JSON-path" at a given position within a YAML document.
 * This function takes the list of lines of the document.
 * The cursor is considered to be after the last character of the last line.
 *
 * The return value is an array of path segments found in order, e.g. [inspectit,config,http].
 * If the path lies within a list, the list index is also correctly added to the path.
 *
 * This function does fail for JSON-style Maps and lists within the yaml.
 * @param {} allLines the lines of the YAML document
 */
export const getYamlPath = (allLines) => {
  let lines = allLines;
  if (findNonQuotedChar(lines[lines.length - 1], '#') !== -1) {
    return null; // we are in a comment
  }

  lines = removeCommentsAndBlankLines(lines);

  let path = beginPath(lines[lines.length - 1]);

  let listIndices = [];

  const lastLineIndentation = getIndentation(lines[lines.length - 1]);
  let parentLineNumber = getNonListParent(lastLineIndentation - 1, lines, lines.length - 1, listIndices);
  path = listIndices.concat(path);

  while (parentLineNumber !== -1) {
    const indentation = getIndentation(lines[parentLineNumber]);
    const statement = decodeStatment(lines[parentLineNumber].substring(indentation));
    if (statement.value) {
      return null;
    }
    path = [statement.key, ...path];
    listIndices = [];
    parentLineNumber = getNonListParent(indentation - 1, lines, parentLineNumber, listIndices);
    path = listIndices.concat(path);
  }
  return path;
};

/**
 *
 * Begins to find the configuration path based on the line which contains the cursor
 * For example, if the line is `    "enabled" : tr`,
 * the returned path is [enabled,tr].
 * Leading indentation of the line is ignored.
 *
 * @param {*} cursorLine
 */
const beginPath = (cursorLine) => {
  const indentation = getIndentation(cursorLine);
  const lastStatement = decodeStatment(cursorLine.substring(indentation));
  if (lastStatement.value) {
    return [lastStatement.key, lastStatement.value];
  } else {
    return [lastStatement.key];
  }
};

/**
 * Removes yaml comments (starting with '#') and all empty lines.
 * If the last line is a blank line, it is kept.
 * @param {*} lines the array of lines
 */
const removeCommentsAndBlankLines = (lines) => {
  lines = lines.map((line) => {
    const idx = findNonQuotedChar(line, '#');
    return idx !== -1 ? line.substring(0, idx) : line;
  });
  // remove all empty lines (except for the last one)
  const lastLine = lines[lines.length - 1];
  lines = lines.filter((line) => line === lastLine || !line.match(/^\s*$/));
  return lines;
};

/**
 * Scans for the parent line of a given starting line in yaml.
 * A parent line is a line in the form `<key>:` which has lower indentation than the starting line.
 * If any dashes are encountered while scannign for the parent line, the resulting list indices are added to the provided
 * "indices" array.
 *
 * For example, given the followign snippet:
 *      config:
 *        - first
 *        - - something
 *          - something else
 *          - last
 * If we invoke this method scanning for the parent pointing to the last line, the result will be the line number of "config:".
 * The indices array will be [1,2].
 * The reason is, that in parsed form config[1][2] will refer to "last".
 *
 * Returns the index of the found parent line or "-1" if no parent exists.
 *
 * @param {*} pos  the index of the last character in the starting line which is part of the indentation.
 * @param {*} lines the array of all lines
 * @param {*} lineNumber the index of the starting line number
 * @param {*} indices the array to place found list-indices within.
 */
const getNonListParent = (pos, lines, lineNumber, indices) => {
  if (pos === -1) {
    return -1;
  }
  const line = lines[lineNumber];
  const firstDash = getFirstDash(line);
  const lastDash = getLastDash(line, pos);
  if (firstDash !== lastDash) {
    const result = getNonListParent(lastDash - 1, lines, lineNumber, indices);
    indices.push(0);
    return result;
  }

  if (firstDash === -1) {
    // no more dashes
    let parentLine = lineNumber - 1;
    const leadingSpaces = pos + 1;
    while (parentLine >= 0 && getLeadingSpaces(lines[parentLine]) >= leadingSpaces) {
      parentLine--;
    }
    if (parentLine !== -1) {
      const parentIndent = getIndentation(lines[parentLine]);
      if (parentIndent >= leadingSpaces) {
        return getNonListParent(leadingSpaces - 1, lines, parentLine, indices);
      }
    }
    return parentLine;
  } else {
    const leadingSpaces = firstDash;
    let parentLine = lineNumber - 1;
    let listIndex = 0;
    while (parentLine >= 0 && (getLeadingSpaces(lines[parentLine]) > leadingSpaces || getFirstDash(lines[parentLine]) === firstDash)) {
      if (lines[parentLine].length > firstDash && lines[parentLine][firstDash] === '-') {
        listIndex++;
      }
      parentLine--;
    }
    if (parentLine >= 0 && lines[parentLine].length > firstDash && lines[parentLine][firstDash] === '-') {
      listIndex++;
    }
    let result = parentLine;
    if (parentLine !== -1) {
      const indent = firstDash + 1;
      const parentIndent = getIndentation(lines[parentLine]);
      if (parentIndent >= indent) {
        result = getNonListParent(firstDash - 1, lines, parentLine, indices);
      }
    }
    indices.push(listIndex);
    return result;
  }
};

/**
 * Extracts the key and the value from a colon-assignment in yaml.
 * Hereby quotes and brackets are removed.
 * Examples.
 *  ` hello: world  ` -> (key = hello, value = world)
 *  ` '[hello]': "world"  ` -> (key = hello, value = world)
 *  ` "hello": "wor  ` -> (key = hello, value = wor)
 *  ` "hello":   ` -> (key = hello, value = <empty string>)
 *  ` "hello"   ` -> (key = hello, value = null)
 *
 * @param {*} statement the input statement as a string
 */
const decodeStatment = (statement) => {
  const colonIndex = findNonQuotedChar(statement, ':');
  let key = unquote(colonIndex !== -1 ? statement.substring(0, colonIndex).trim() : statement.trim());
  const value = colonIndex !== -1 ? unquote(statement.substring(colonIndex + 1).trim()) : null;
  key = removeBrackets(key);

  return { key, value };
};

/**
 * Removes quotes from the beginnin and end of a string.
 * Leading quotes (single and double) are always removed.
 * Trailing quotes are only removed if a leading quote was removed.
 *
 * @param {*} text the string to remove the quotes from.
 */
const unquote = (text) => {
  if (text.startsWith("'") || text.startsWith('"')) {
    if (text.endsWith("'") || text.endsWith('"')) {
      return text.substring(1, text.length - 1);
    } else {
      return text.substring(1, text.length);
    }
  }
  return text;
};

/**
 * Removes Brackets "[" and "]" from the beginnin and end of a string.
 * Leading "[" are always removed.
 * Trailing "]" are only removed if a leading "[" was removed.
 *
 * @param {*} text the string to remove the brackets from.
 */
const removeBrackets = (text) => {
  if (text.startsWith('[')) {
    if (text.endsWith(']')) {
      return text.substring(1, text.length - 1);
    } else {
      return text.substring(1, text.length);
    }
  }
  return text;
};

/**
 * Finds the first occurence of a given character in a given string.
 * Hereby occurences are ignored which are within single or double quotes.
 *
 * Returns the index of the found occurece or -1 if none is found.
 *
 * @param {*} str the string to scan within
 * @param {*} char the character to scan for
 */
const findNonQuotedChar = (str, char) => {
  const NORMAL = 0;
  const SINGLE_QUOTED = 1;
  const DOUBLE_QUOTED = 2;
  let state = NORMAL;
  for (let i = 0; i < str.length; i++) {
    if (str[i] === char && state === NORMAL) {
      return i;
    } else if (str[i] === "'") {
      if (state === NORMAL) {
        state = SINGLE_QUOTED;
      } else if (state === SINGLE_QUOTED) {
        state = NORMAL;
      }
    } else if (str[i] === '"') {
      if (state === NORMAL) {
        state = DOUBLE_QUOTED;
      } else if (state === DOUBLE_QUOTED) {
        state = NORMAL;
      }
    }
  }
  return -1;
};

/**
 * Returns the number of leading spaces or dashes for the given string.
 *
 * @param {*} line the string to get the indentation of
 */
const getIndentation = (line) => {
  let pos = 0;
  while (pos < line.length && (line[pos] === ' ' || line[pos] === '-')) {
    pos++;
  }
  return pos;
};

/**
 * Returns the number of leading spaces for the given string.
 *
 * @param {*} line the string to get the number of leading spaces from
 */
const getLeadingSpaces = (line) => {
  let pos = 0;
  while (pos < line.length && line[pos] === ' ') {
    pos++;
  }
  return pos;
};

/**
 * Returns the index of the first dash after spaces for the given string.
 * Returns -1 if there is no dash found.
 *
 * @param {*} line the string to find the first dash within
 */
const getFirstDash = (line) => {
  let pos = 0;
  while (pos < line.length && line[pos] === ' ') {
    pos++;
  }
  if (pos < line.length && line[pos] === '-') {
    return pos;
  } else {
    return -1;
  }
};

/**
 * Scans backwards in a given string, starting at a specified index, for a dash.
 * If a dash is found, it's index is returned.
 * Otherwise -1 is returned.
 *
 * @param {*} line the string to find the first dash within
 * @param {*} posInclusive the index to start scanning at
 */
const getLastDash = (line, posInclusive) => {
  let pos = posInclusive;
  while (pos >= 0 && line[pos] !== '-') {
    pos--;
  }
  return pos;
};
