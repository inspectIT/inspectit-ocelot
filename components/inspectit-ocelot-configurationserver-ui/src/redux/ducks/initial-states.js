import SidebarTypes from '../../components/views/configuration/SidebarTypes';

const authentication = {
  /** The access token which is used for bearer authentication. */
  token: null,
  /** The authorization permissions the user has*/
  permissions: {
    write: false,
    promote: false,
    admin: false,
  },
  /** Specifying whether a login request is currently be executed */
  loading: false,
  /** This contains the error message if the last request has failed. */
  error: null,
  /** The username of the currently logged in user. */
  username: null,
};

const configuration = {
  /** Specifies how many requests are currently loading in the background */
  pendingRequests: 0,
  /** The existing versions. */
  versions: [],
  /** The existing configuration files. */
  files: [],
  /** Indicates if hidden files ares shown. */
  showHiddenFiles: false,
  /**
   * The history of file-move operations which were executed successfully.
   * When the files are refetched, it can occur that the selection points to a non existing file.
   * Normally, the selection is then simply removed.
   * With the moveHistory, it is possible to corret the selection, in case the selected file was moved.
   * The moveHistory is cleared after a files have been refetched.
   */
  moveHistory: [
    /* {source, target} */
  ],
  /** The date when the configuration files have been fetched. */
  updateDate: null,
  /** The (abosolute) path of the currently selected file in the configuration file tree. */
  selection: null,
  /** The content of the currently selected file. */
  selectedFileContent: null,
  /**
   * A map of unsaved file changes.
   * Maps the absolute file path to the files contents.
   */
  unsavedFileContents: {
    /* fileName: fileContents */
  },
  /** The default configuration of the Ocelot agents. Will be retrieved as key/value pairs each representing path/content of a file. */
  defaultConfig: {},
  /** The path of the currently selected default configuration file in the file tree. */
  selectedDefaultConfigFile: null,
  /** The configuration schema */
  schema: null,
  /** If the config view should be in the visual mode */
  showVisualConfigurationView: false,
  /** Specifies the selected git version. The latest version always has the number 0. */
  selectedVersion: null,
  /** Defines which sidebar is currently shown */
  currentSidebar: SidebarTypes.NONE,
};

const notification = {
  /** The latest notification object. */
  lastNotification: null,
};

const mappings = {
  /** Specifies how many requests are currently loading in the background. */
  pendingRequests: 0,
  /** The current agent mappings. */
  mappings: [],
  /** The date when the agent mappings have been fetched. */
  updateDate: null,
  /** The existing versions. */
  versions: [],
  /** Specifies the selected git version. The latest version always has the number 0. */
  selectedVersion: null,
  /** The source branch for the agent mappings file itself. */
  sourceBranch: '',
  /** Defines which sidebar is currently shown */
  currentSidebar: SidebarTypes.NONE,
};

const alerting = {
  /** A mapping of rule names and corresponding unsaved contents.*/
  unsavedRuleContents: {},
  /** A mapping of handler ids and corresponding unsaved contents. */
  unsavedHandlerContents: {},
  ruleGrouping: {
    groupByTemplates: true,
    groupByTopics: false,
  },
};

const agentStatus = {
  /**
   * The list of connected agents, each agent is an object with the following structure:
   * {attributes: (dictionary), mappingName: string, lastConfigFetch: timestamp}
   */
  agents: [],
  /** Specifies how many requests are currently loading the data in the background */
  pendingRequests: 0,
  /** Specifies how many requests are currently clearing the data in the background */
  pendingClearRequests: 0,
};

const settings = {
  /** The list of users */
  users: [],
  /** Specifies how many requests are currently loading the data in the background */
  pendingRequests: 0,
};

const promotion = {
  /** The names of the files which are currently be approved */
  approvals: [],
  /** The commit id of the working directory related to the currently changed files */
  workspaceCommitId: null,
  /** The commit id of the live directory related to the currently changed files */
  liveCommitId: null,
};

export { authentication, configuration, notification, mappings, agentStatus, settings, promotion, alerting };
