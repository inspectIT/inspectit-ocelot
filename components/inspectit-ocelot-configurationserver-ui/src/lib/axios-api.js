import axios from 'axios';
import { getStore } from './with-redux-store';
import { authenticationActions } from '../redux/ducks/authentication';
import { notificationActions } from '../redux/ducks/notification';

const isProduction = process.env.NODE_ENV === 'production';
const BASE_API_URL_V1 = isProduction ? '/api/v1' : 'http://localhost:8090/api/v1';

const commonConfiguration = {
  baseURL: BASE_API_URL_V1,
  timeout: 15000,
};

const axiosPlain = axios.create(commonConfiguration);
const axiosBearer = axios.create(commonConfiguration);

// ############################################################################
// Request Interceptors

/**
 * Ensures requres are authenticated using the bearer token.
 */
axiosBearer.interceptors.request.use(
  function (config) {
    const state = getStore().getState();
    const { token } = state.authentication;

    config.headers.authorization = 'Bearer ' + token;

    return config;
  },
  function (error) {
    return Promise.reject(error);
  }
);

// ############################################################################
// Response Interceptors

/**
 * Does error handling. Shows notifications in case an error occurs or triggers
 * a logout if a request returns with a 401 status code (unauthorized).
 */
axiosBearer.interceptors.response.use(
  function (response) {
    return response;
  },
  function (error) {
    const reduxStore = getStore();

    if (error.response && error.response.status === 401) {
      reduxStore.dispatch(
        notificationActions.showWarningMessage('Unauthorized', 'Your access token is no longer valid. Please login again.')
      );
      reduxStore.dispatch(authenticationActions.logout());
    } else if (error.response && error.response.data && error.response.data.message) {
      const { message } = error.response.data;
      reduxStore.dispatch(notificationActions.showErrorMessage('Request failed', message));
    } else {
      const { message } = error;
      reduxStore.dispatch(notificationActions.showErrorMessage('Request failed', message));
    }

    return Promise.reject(error);
  }
);

export { axiosPlain };

export default axiosBearer;
