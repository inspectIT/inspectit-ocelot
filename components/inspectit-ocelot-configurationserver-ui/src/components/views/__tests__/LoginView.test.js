import { render, screen } from '@testing-library/react';
import LoginView from '../LoginView';
import '@testing-library/jest-dom';
import React from 'react';
import { storeWrapper } from '../../../lib/reduxTestUtils';
import { authentication } from '../../../redux/ducks';
import userEvent from '@testing-library/user-event';

const setup = () => {
  const reducers = { authentication };
  return render(storeWrapper(<LoginView />, reducers));
};

describe('LoginView', () => {
  let container;
  //Arrange
  beforeEach(() => {
    const renderedDom = setup();
    container = renderedDom.container;
  });

  it('renders successfully', () => {
    expect(container).toMatchSnapshot();
  });

  it('disables the login button when username is missing', async () => {
    //Arrange
    const loginButton = screen.getByRole('button', { name: 'Login' });
    const password = screen.getByPlaceholderText('Password');

    //Act
    await userEvent.type(password, 'password123');

    //Assert
    expect(loginButton).toBeDisabled();
  });

  it('disables the login button when password is missing', async () => {
    //Arrange
    const loginButton = screen.getByRole('button', { name: 'Login' });
    const username = screen.getByPlaceholderText('Username');

    //Act
    await userEvent.type(username, 'userName');

    //Assert
    expect(loginButton).toBeDisabled();
  });

  it('enables the login button when username and password are present', async () => {
    //Arrange
    const loginButton = screen.getByRole('button', { name: 'Login' });
    const username = screen.getByPlaceholderText('Username');
    const password = screen.getByPlaceholderText('Password');

    //Act
    await userEvent.type(username, 'userName');
    await userEvent.type(password, 'password123');

    //Assert
    expect(loginButton).toBeEnabled();
  });
});
