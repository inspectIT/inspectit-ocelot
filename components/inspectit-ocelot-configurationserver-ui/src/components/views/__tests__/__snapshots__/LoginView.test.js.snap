// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginView renders successfully 1`] = `
<div>
  <div
    class="jsx-3064013252 this"
  >
    <div
      class="jsx-128945688 this"
    >
      <div
        class="jsx-3380344897 this"
      >
        <img
          class="jsx-3380344897 ocelot-head"
          src="/static/images/inspectit-ocelot.svg"
        />
        <div
          class="jsx-3380344897 text-ocelot"
        >
          inspectIT Ocelot
        </div>
        <div
          class="jsx-3380344897 text-server"
        >
          Configuration Server
        </div>
      </div>
      <div
        class="jsx-128945688 p-inputgroup input"
      >
        <span
          class="jsx-128945688 p-inputgroup-addon"
        >
          <i
            class="jsx-128945688 pi pi-user"
          />
        </span>
        <input
          class="p-inputtext p-component"
          placeholder="Username"
          style="width: 100%;"
          value=""
        />
      </div>
      <div
        class="jsx-128945688 p-inputgroup input"
      >
        <span
          class="jsx-128945688 p-inputgroup-addon"
        >
          <i
            class="jsx-128945688 pi pi-lock"
          />
        </span>
        <input
          class="p-inputtext p-component"
          placeholder="Password"
          style="width: 100%;"
          type="password"
          value=""
        />
      </div>
      <div
        class="jsx-128945688 input"
      >
        <button
          class="p-button p-component p-button-text-only p-disabled"
          disabled=""
          style="width: 100%;"
        >
          <span
            class="p-button-text p-c"
          >
            Login
          </span>
        </button>
      </div>
    </div>
    <div
      class="jsx-3064013252 build-information"
    >
      <span
        class="jsx-3064013252"
      >
        inspectIT Ocelot Configuration Server v
         (Build Date: 
        ) |
         
        <a
          class="jsx-3064013252"
          href="http://docs.inspectit.rocks/"
          target="_blank"
        >
          Docs
        </a>
         
        |
         
        <a
          class="jsx-3064013252"
          href="https://github.com/inspectIT/inspectit-ocelot"
          target="_blank"
        >
          Github
        </a>
      </span>
    </div>
  </div>
</div>
`;
